File: dataset/output/PERPLEXITY_llama-3-sonar-large-32k-chat/Task92_PERPLEXITY_llama-3-sonar-large-32k-chat.cpp
#include <iostream>
#include <thread>
#include <mutex>

std::mutex lock;
int maxCounter = 5;
int currentCounter = 0;

void accessResource(int threadId) {
    while (currentCounter <= maxCounter) {
        std::lock_guard<std::mutex> guard(lock);
        if (currentCounter <= maxCounter) {
            currentCounter++;
            std::cout << "Thread " << threadId << " is accessing the currentCounter at " << currentCounter << std::endl;
        } else {
            break;
        }
    }
}

int main() {
    std::thread threads[5];
    for (int i = 0; i < 5; i++) {
        threads[i] = std::thread(accessResource, i);
    }
    for (auto& t : threads) {
        t.join();
    }
    return 0;
}

File: dataset/output/PERPLEXITY_llama-3-sonar-large-32k-chat/task92/Task92_PERPLEXITY_llama-3-sonar-large-32k-chat.java
class SharedResource {
    int maxCounter;
    int currentCounter;
    Object lock = new Object();

    public SharedResource(int maxCounter) {
        this.maxCounter = maxCounter;
        this.currentCounter = 0;
    }

    public void accessResource(String threadName) {
        while (currentCounter <= maxCounter) {
            synchronized (lock) {
                if (currentCounter <= maxCounter) {
                    currentCounter++;
                    System.out.println("Thread " + threadName + " is accessing the currentCounter at " + currentCounter);
                } else {
                    break;
                }
            }
        }
    }

    public static void main(String[] args) {
        SharedResource resource = new SharedResource(5);
        Thread[] threads = new Thread[5];
        for (int i = 0; i < 5; i++) {
            threads[i] = new Thread(() -> resource.accessResource(String.valueOf(i)));
            threads[i].start();
        }
        for (Thread t : threads) {
            try {
                t.join();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }
}